pkg/mimir/tracing.go

```diff
@@ -36,3 +42,195 @@ type wrappedServerStream struct {
+		}
+	}
+
+	span, ctx := opentracing.StartSpanFromContextWithTracer(ctx, t.tracer, spanName, mappedOptions...)
+	return ctx, NewOpenTelemetrySpanBridge(span, t.provider)
+}
+
+type OpenTelemetrySpanBridge struct {
+	span     opentracing.Span
+	provider trace.TracerProvider
+}
+
+func NewOpenTelemetrySpanBridge(span opentracing.Span, provider trace.TracerProvider) *OpenTelemetrySpanBridge {
+	return &OpenTelemetrySpanBridge{
+		span:     span,
+		provider: provider,
+	}
+}
+
+// End completes the Span. The Span is considered complete and ready to be
+// delivered through the rest of the telemetry pipeline after this method
+// is called. Therefore, updates to the Span are not allowed after this
+// method has been called.
+func (s *OpenTelemetrySpanBridge) End(options ...trace.SpanEndOption) {
+	if len(options) == 0 {
+		s.span.Finish()
+		return
+	}
+
+	cfg := trace.NewSpanEndConfig(options...)
+	s.span.FinishWithOptions(opentracing.FinishOptions{
+		FinishTime: cfg.Timestamp(),
+	})
+}
+
+// AddEvent adds an event with the provided name and options.
+func (s *OpenTelemetrySpanBridge) AddEvent(name string, options ...trace.EventOption) {
+	// Options are not supported.
+	s.span.LogFields(log.Event(name))
+}
+
+// IsRecording returns the recording state of the Span. It will return
+// true if the Span is active and events can be recorded.
+func (s *OpenTelemetrySpanBridge) IsRecording() bool {
+	return true
+}
+
+// RecordError will record err as an exception span event for this span. An
+// additional call to SetStatus is required if the Status of the Span should
+// be set to Error, as this method does not change the Span status. If this
+// span is not being recorded or err is nil then this method does nothing.
+func (s *OpenTelemetrySpanBridge) RecordError(err error, options ...trace.EventOption) {
+	// Options are not supported.
+	s.span.LogFields(log.Error(err))
+}
+
+// SpanContext returns the SpanContext of the Span. The returned SpanContext
+// is usable even after the End method has been called for the Span.
+func (s *OpenTelemetrySpanBridge) SpanContext() trace.SpanContext {
+	// We only support Jaeger span context.
+	sctx, ok := s.span.Context().(jaeger.SpanContext)
+	if !ok {
+		return trace.SpanContext{}
+	}
+
+	var flags trace.TraceFlags
+	flags = flags.WithSampled(sctx.IsSampled())
+
+	return trace.NewSpanContext(trace.SpanContextConfig{
+		TraceID:    jaegerToOpenTelemetryTraceID(sctx.TraceID()),
+		SpanID:     jaegerToOpenTelemetrySpanID(sctx.SpanID()),
+		TraceFlags: flags,
+
+		// Unsupported because we can't read it from the Jaeger span context.
+		Remote: false,
+	})
+}
+
+// SetStatus sets the status of the Span in the form of a code and a
+// description, overriding previous values set. The description is only
+// included in a status when the code is for an error.
+func (s *OpenTelemetrySpanBridge) SetStatus(code codes.Code, description string) {
+	// Not supported.
```

Comment:
This is used in prometheus for example to indicate that the span had an error ([example](https://github.com/grafana/mimir/blob/8fca3461c1e9f74cdbc98a8fc1fcfbbdcbbe24da/vendor/github.com/prometheus/prometheus/rules/manager.go#L641)). I think the result is that it shows as a red alert sign in UIs. We can:
```suggestion
	s.span.SetTag(code.String(), description)
```

Reply:
> A few questions about some features that weren't ported. I am not sure how well these bridges should translate between OTEL and OT.

Thanks for the valuable review! The bridge we need should be "good enough", given the limited real usage we're doing right now. Anyway, feedback was good!

Reply:
The red alert sign in the UI is displayed when you log an errr, not when you set a tag:
![Screenshot 2022-08-12 at 15 41 28](https://user-images.githubusercontent.com/1701904/184365967-989ff70c-1c07-498c-aae1-03d2b2dba0da.png)